bundles/binding/org.openhab.binding.stiebelheatpump/src/main/java/org/openhab/binding/stiebelheatpump/internal/StiebelHeatPumpBinding.java

```diff
@@ -0,0 +1,389 @@
+		ManagedService {
+
+	private static final Logger logger = LoggerFactory
+			.getLogger(StiebelHeatPumpBinding.class);
+
+	// configuration defaults for optional properties
+	private static int DEFAULT_BAUD_RATE = 9600;
+	private static int DEFAULT_SERIAL_TIMEOUT = 5;
+
+	/**
+	 * the refresh interval which is used to poll values from the heat pump
+	 * server (optional, defaults to 1 Minute)
+	 */
+	private long refreshInterval = 60000; // in ms
+
+	/** the serial port to use for connecting to the heat pump */
+	private String serialPort;
+
+	/** baud rate for the serial port */
+	private int baudRate = DEFAULT_BAUD_RATE;
+
+	/** timeout for the serial port */
+	private int serialTimeout = DEFAULT_SERIAL_TIMEOUT;
+
+	/** version number of heat pump */
+	private String version = "";
+
+	/** heat pump request definition */
+	private List<Request> heatPumpConfiguration = new ArrayList<Request>();
+
+	// ** request to get the version of the heat pump
+	Request versionRequest;
+
+	// ** indicates if the communication is currently in use by a call
+	boolean communicationInUse = false;
+
+	public StiebelHeatPumpBinding() {
+	}
+
+	public void activate() {
+
+	}
+
+	public void deactivate() {
+
+	}
+
+	/**
+	 * @{inheritDoc
+	 */
+	protected long getRefreshInterval() {
+		return refreshInterval;
+	}
+
+	/**
+	 * @{inheritDoc
+	 */
+	protected String getName() {
+		return "stiebelheatpump Refresh Service";
+	}
+
+	/**
+	 * @{inheritDoc
+	 */
+	protected void execute() {
+
+		if (communicationInUse)
+			return;
+
+		logger.debug("Refresh heat pump sensor and status values ...");
+		communicationInUse = true;
+		CommunicationService communicationService = null;
+
+		try {
+			communicationService = new CommunicationService(serialPort,
+					baudRate, heatPumpConfiguration);
+			Map<String, String> data = new HashMap<String, String>();
+			Map<String, String> allData = new HashMap<String, String>();
+
+			data = communicationService.getStatus();
+			allData.putAll(data);
+			for (Map.Entry<String, String> entry : data.entrySet()) {
+				logger.debug("Data {} has value {}", entry.getKey(),
+						entry.getValue());
+			}
+			data = communicationService.getSensors();
+			allData.putAll(data);
+			for (Map.Entry<String, String> entry : data.entrySet()) {
+				logger.debug("Data {} has value {}", entry.getKey(),
+						entry.getValue());
+			}
+
+			communicationService.finalizer();
+
+			publishValues(allData);
+		} catch (StiebelHeatPumpException e) {
+			logger.error("Could not read data from heat pump! " + e.toString());
+		} finally {
+			communicationService.finalizer();
+			communicationInUse = false;
+		}
+	}
+
+	/**
+	 * @{inheritDoc
+	 */
+	protected void internalReceiveCommand(String itemName, Command command) {
+		logger.debug("Received command {} for item {}", command, itemName);
+
+		int retry = 0;
+		while (communicationInUse & (retry < DEFAULT_SERIAL_TIMEOUT)) {
+			try {
+				Thread.sleep(CommunicationService.WAITING_TIME_BETWEEN_REQUESTS);

```

Comment:
letting the Thread sleep here slows down the overall openHAB event processing since this is still single threaded